// cucumber/lib/formatter/helpers/summary_helpers.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatSummary = formatSummary;

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _duration = require('duration');

var _duration2 = _interopRequireDefault(_duration);

var _status = require('../../status');

var _status2 = _interopRequireDefault(_status);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var STATUS_REPORT_ORDER = [_status2.default.FAILED, _status2.default.AMBIGUOUS, _status2.default.UNDEFINED, _status2.default.PENDING, _status2.default.SKIPPED, _status2.default.PASSED];

function formatSummary(_ref) {
  var colorFns = _ref.colorFns,
      testCaseMap = _ref.testCaseMap,
      testRun = _ref.testRun;

  var testCaseResults = [];
  var testStepResults = [];
  _lodash2.default.each(testCaseMap, function (_ref2) {
    var result = _ref2.result,
        steps = _ref2.steps;

    testCaseResults.push(result);
    _lodash2.default.each(steps, function (testStep) {
      if (testStep.sourceLocation) {
        testStepResults.push(testStep.result);
      }
    });
  });
  var scenarioSummary = getCountSummary({
    colorFns: colorFns,
    objects: testCaseResults,
    type: 'scenario'
  });
  var stepSummary = getCountSummary({
    colorFns: colorFns,
    objects: testStepResults,
    type: 'step'
  });
  var durationSummary = getDuration(testRun.result.duration);
  return [scenarioSummary, stepSummary, durationSummary].join('\n');
}

function getCountSummary(_ref3) {
  var colorFns = _ref3.colorFns,
      objects = _ref3.objects,
      type = _ref3.type;

  var counts = _lodash2.default.chain(objects).groupBy('status').mapValues('length').value();
  var total = _lodash2.default.reduce(counts, function (memo, value) {
    return memo + value;
  }) || 0;
  var text = total + ' ' + type + (total === 1 ? '' : 's');
  if (total > 0) {
    var details = [];
    STATUS_REPORT_ORDER.forEach(function (status) {
      if (counts[status] > 0) {
        details.push(colorFns[status](counts[status] + ' ' + status));
      }
    });
    text += ' (' + details.join(', ') + ')';
  }
  return text;
}

function getDuration(milliseconds) {
  var start = new Date(0);
  var end = new Date(milliseconds);
  var duration = new _duration2.default(start, end);

  return duration.minutes + 'm' + duration.toString('%S') + '.' + duration.toString('%L') + 's' + '\n';
}
