// cucumber/lib/support_code_library/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SupportCodeLibraryBuilder = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _parameter_type_registry_builder = require('./parameter_type_registry_builder');

var _parameter_type_registry_builder2 = _interopRequireDefault(_parameter_type_registry_builder);

var _define_helpers = require('./define_helpers');

var _finalize_helpers = require('./finalize_helpers');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SupportCodeLibraryBuilder = exports.SupportCodeLibraryBuilder = function () {
  function SupportCodeLibraryBuilder() {
    var _this = this;

    (0, _classCallCheck3.default)(this, SupportCodeLibraryBuilder);

    this.methods = {
      addTransform: (0, _define_helpers.addTransform)(this),
      defineParameterType: (0, _define_helpers.defineParameterType)(this),
      After: (0, _define_helpers.defineScenarioHook)(this, 'afterScenarioHookDefinitions'),
      AfterAll: (0, _define_helpers.defineFeaturesHook)(this, 'afterFeaturesHookDefinitions'),
      Before: (0, _define_helpers.defineScenarioHook)(this, 'beforeScenarioHookDefinitions'),
      BeforeAll: (0, _define_helpers.defineFeaturesHook)(this, 'beforeFeaturesHookDefinitions'),
      defineSupportCode: function defineSupportCode(fn) {
        fn(_this.methods());
      },
      defineStep: (0, _define_helpers.defineStep)(this),
      registerHandler: (0, _define_helpers.registerHandler)(this),
      registerListener: function registerListener(listener) {
        _this.options.listeners.push(listener);
      },
      setDefaultTimeout: function setDefaultTimeout(milliseconds) {
        _this.options.defaultTimeout = milliseconds;
      },
      setDefinitionFunctionWrapper: function setDefinitionFunctionWrapper(fn) {
        _this.options.definitionFunctionWrapper = fn;
      },
      setWorldConstructor: function setWorldConstructor(fn) {
        _this.options.World = fn;
      }
    };
    this.methods.Given = this.methods.When = this.methods.Then = this.methods.defineStep;
  }

  (0, _createClass3.default)(SupportCodeLibraryBuilder, [{
    key: 'finalize',
    value: function finalize() {
      var _this2 = this;

      (0, _finalize_helpers.wrapDefinitions)({
        cwd: this.cwd,
        definitionFunctionWrapper: this.definitionFunctionWrapper,
        definitions: _lodash2.default.chain(['afterFeaturesHook', 'afterScenarioHook', 'beforeFeaturesHook', 'beforeScenarioHook', 'step']).map(function (key) {
          return _this2.options[key + 'Definitions'];
        }).flatten().value()
      });
      this.options.afterScenarioHookDefinitions.reverse();
      this.options.afterFeaturesHookDefinitions.reverse();
      return this.options;
    }
  }, {
    key: 'reset',
    value: function reset(cwd) {
      this.cwd = cwd;
      this.options = _lodash2.default.cloneDeep({
        afterFeaturesHookDefinitions: [],
        afterScenarioHookDefinitions: [],
        beforeFeaturesHookDefinitions: [],
        beforeScenarioHookDefinitions: [],
        defaultTimeout: 5000,
        definitionFunctionWrapper: null,
        listeners: [],
        stepDefinitions: [],
        parameterTypeRegistry: _parameter_type_registry_builder2.default.build(),
        World: function World(_ref) {
          var attach = _ref.attach,
              parameters = _ref.parameters;

          this.attach = attach;
          this.parameters = parameters;
        }
      });
    }
  }]);
  return SupportCodeLibraryBuilder;
}();

exports.default = new SupportCodeLibraryBuilder();
